Refactor Transitions and Moves

Adda0 opened the pull request:
This PR refactors Moves and Transitions according to what we have agreed on. Namely, it:
- refactor transition and move iterators into a simple single class objects with their own const iterators. 
- includes epsilon and symbol move iterators.
- updates Python binding to work with the current implementation over Transitions.
- moves methods about iterating over transitions into Transitions.

Questions to resolve:
- Do we need `Delta::transitions_begin()` (`..._end()`) and `StatePost::moves_begin()` (`epsilon_moves_...`, `symbol_moves_...`; `..._end()`) methods? Currently, you have to get the `Moves` and `Transitions` classes to get begin or end iterator to moves and transitions.

Adda0 committed:
Add instance of 'Transitions' as an owned member of 'Delta'

Adda0 committed:
Move get_transitions_from_as_sequence() to Delta::Transitions::from()

Adda0 committed:
Update Python binding to use transitions iterators

Adda0 committed:
Check corrent functionality of Delta::operator=

Adda0 committed:
Remove iterator over Delta

Adda0 committed:
Compare Delta by their transitions as 'Transition'

Adda0 committed:
Refactor moves

Adda0 committed:
Move getting number of transitions to Delta

Adda0 committed:
Move const iterator over transitions into Transitions class

Adda0 committed:
Remove getting transtions as sequence

Adda0 committed:
Refactor Moves to reflect structure of Transitions

Adda0 committed:
Update Python binding to new moves and transitions implemenentation

Adda0 committed:
Set end to true for default const iterators over moves and transitions

Adda0 committed:
Add support for iterating over epsilon moves and symbol (non-epsilon) moves

Adda0 commented:
@vhavlena Since you have been invested in the discussion about transitions and moves, could you have a look at how the transitions and moves were refactored (what is the current interface for them) and give your opinion on how you are satisfied with it?

vhavlena commented:
It seems fine. I would rename `const_iterator` to `ConstIterator` to match the convention.

kilohsakul commented on src/nfa/delta.cc:
```diff
@@ -59,12 +59,9 @@
-        for (const SymbolPost & m: (*this)[q]) { size = size + m.size(); }
-    }
-    return size;
+Delta& Delta::operator=(const Delta& other) {
```
=default??

Adda0 replied:
It could not be default implemented when we had transitions as a member of Delta. When we no longer have that, it can be default implemented again using the member-wise copy assignment.

Adda0 replied:
Fixed.

tfiedor commented on include/mata/nfa/delta.hh:
```diff
@@ -121,64 +121,105 @@ public:
+
+         * Iterates over symbol posts with symbols in interval [first_symbol_, last_symbol_].
+         */
+        enum class FirstSymbolLookupDiretion {
```
```suggestion
        enum class FirstSymbolLookupDirection {
```
Maybe `LookupDirection` would suffice? Or `IterationDirection`?

Adda0 replied:
It is not an iteration direction. The moves are still iterated in the order of first to last, but the iteration starts at some first symbol which must be found in the vector of `SymbolPost`s first. The process of finding the iterator to this first symbol (and its `SymbolPost`) can proceed either backward (from the last element in the vector to the first one) or forward (from the first element to the last one). When you are looking for the first epsilon in the vector, you want to start from the end. Otherwise, you should start from the beginning of the vector.

tfiedor replied:
OK.

kilohsakul replied:
We discussed and agreed to change it, this class will be gone, but I forgot what exactly should happen.

Adda0 replied:
We will have only one function, `StatePost::moves()` which will take 3 optional parameters. Without any, it will behave as a normal moves iterator iterating over all moves (including epsilons) and when the parameters are specified, it will iterate over some subset of these: `StatePost::moves(moves_kind, first_symbol_iterator, last_symbol_iterator)` where `moves_kind` is of type enum class with values `MovesKind::All`, `MovesKind::Epsilons` and `MovesKind::Symbols`. Iterators then only specify, which subset of moves is to be iterated over. The kind specifies whether to look for the first symbol from the left or from the right. This replaces `FirstSymbolLookupDirection` class.

tfiedor replied:
Sounds much better. Moreover, I though of few alternatives for `MovesKinds` (sounds kinda weird): `Pick::All`, `Select::Epsilons` or maybe `Iterate::Symbols`? (I wrote them with the second part as well, so one can see, how it sounds). Feel free to pick one, if you like any.

Adda0 replied:
I am experimenting with the names and `Iterate::Symbols` seems good to me. I will use `Iterate` and others can give their opinion in the PR when the changes are pushed.

kilohsakul replied:
"Iterate" does not give much hint, no? 
But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

Related to this, we should fix what "symbol" means. Currently we have the data type Symbol, that can be alphabet symbol or epsilon, but here David was using symbol as alphabet symbol only.
We could have symbol, alphabet symbol, epsilon, or perhaps better "Label" (for anything), "Symbol", and "Epsilon".

Adda0 replied:
I think it explains it well: `nfa.delta[4].moves(Iterate::Epsilons)` gives me all the information I need, in my opinion. I personally do not see a difference between `Iterate` and `IterateOver`. But I can change it to `IterateOver` if you think that is better. Either sounds good to me.

Regarding the `Symbol` naming. You said that you do not consider epsilon to be a symbol, and want the “normal” symbols to be called just symbols. Therefore, the name gets a bit ambiguous. I think we can think about possible names, discuss it tomorrow in our meeting and later refactor the whole Mata accordingly. For now, I will use “alphabet symbols” for “normal” symbols and “epsilons” for epsilon symbols.

tfiedor replied:
> "Iterate" does not give much hint, no? But I don't know what it should be, IterateOver? SymbolKind / SymbolType....

I think, here, you have to take into the account the context and the usage. Since, this will be probably called as `*.moves(Iterate::Symbols, ...)`, just simple verb with the noun is IMO fine.

kilohsakul replied:
Hm, maybe one could use it, but is it sounds ilogical to me. If I read *.moves(bla) as a constructor of a set of moves 
(is this right? The set is not explicitly constructed, only iterable, but one still sees it as a set of moves.) 
then the bla should specify a property of the elements I want to include in the set. Iterate is not a porperty of an element, moves(Iterate:bla) would look like constructing the set of all moves with s special iteration that only considers its subset, weird.

tfiedor replied:
Whatever, I was thinking that it is iterator, and hence its parameters would specify properties of the iteration not its elements, but this discussion is getting nowhere, so go for the `MovesKind`.

Adda0 replied:
>  (is this right? The set is not explicitly constructed, only iterable, but one still sees it as a set of moves.)

It is right.

> then the bla should specify a property of the elements I want to include in the set.

I do not agree. It can be just as well a property of the function (the constructor) which specifies more closely what to construct. However, I have no problems with `IteraveOver`. Are you OK with `IterateOver`? `state_post.moves(IterateOver::Epsilons)`?

Adda0 replied:
We agreed now to keep the original three methods to create `Moves`. Therefore, we do not need this parameter to pick which symbol posts we want to iterate over. This parameter will be changed back to `SeekFirstSymbolDirection::Forward` and will be used by the user only when they want to explicitly fill all parameters themselves: `moves(first_symbol, last_symbol, direction)`. Otherwise, in 99 % of cases, they will just use one of the three functions without needing to specify any parameter.

kilohsakul replied:
Wait, what is the SeekFirstSymbolDirection::Forward ? That should be completely gone, no?

Adda0 replied:
You said in our discussion in Discord to rename `Iterate` to `SeekFirst` or something similar to what I had earlier. I can rename it to something else, but the meaning stays. It no longer represents a kind of moves to create. We have named functions for that. It only means where to start looking for the first symbol. What would you prefer?

Alternatively, I can completely remove this optimization seeing that it brings us so much headache and just look for the first symbol using binary search. Only after checking whether the first and last symbol posts in the state post are already what we look for.

Adda0 replied:
The `SeekFirst...` was replaced by a simple boolean flag parameter for constructor of `Moves`.

Adda0 replied:
Moved the logic getting begin and end symbol post iterators to `moves_...()` functions. Therefore, we cannot create a custom `Moves` class instance with a specified range of `[first_symbol, last_symbol]`. We can only create custom `Moves` class instance with iterators. If we want to specify the range with symbols, we need to use one of the helper functions for getting all labels, symbols, or epsilons.

tfiedor commented on include/mata/nfa/delta.hh:
```diff
@@ -121,64 +121,105 @@ public:
+            const StatePost* state_post_{ nullptr };
+            StatePost::const_iterator state_post_it_{};
+            StateSet::const_iterator symbol_post_it_{};
+            Symbol last_symbol_{ Limits::max_symbol };
```
I wonder if `sentinel_` would be appropriate name? It's short, elegant, well-established.

kilohsakul replied:
In act I would not know what sentinel means.

tfiedor replied:
I believe, it is quite well-established naming for our `zarážka` in programming (e.g. in some lookup algorithm, I am fairly sure it is used). `last_symbol` sounds like something at `.back()`, which I'm not sure if applies for this particular case.

Adda0 replied:
To my best knowledge, it is as @tfiedor says. However, in this case I meant to truly use the notion of first symbol and last symbol to explicitly state that we iterate over a closed interval `[fist_symbol, last_symbol]`, that is, both first symbol and last symbol are iterated over. Sentinel is what we usually use for one after the last element, it is basically what we call the end iterator in C++ iterator semantics.

tfiedor replied:
How about just `max_symbol` then?

tfiedor replied:
My issue is, that it is little ambiguous (could be last as in position, or last as last seen, or last as upper limit, etc.)

Adda0 replied:
`max_symbol` is fine with me. When talking about intervals, you use `[first, last]`, however, instead of `[first, max]`, no?

But I agree that in order to avoid ambiguous code, `max` might be easier to work with.

jurajsic replied:
For me `max` is more ambigious than `last`, especially when it is initialized to `Limits::max_symbol`. I think a comment here would be good.

Adda0 replied:
So what about keeping `last_symbol`, but adding an explanation, that it is by default the maximal symbol and that it means the last symbol in the vector, not the last + 1. Would that be sufficient?

Adda0 replied:
I added more explanations. Hopefully, the distinction will now be clear.

tfiedor reviewed (approved):
No Major nor Colonel Issues.

Adda0 committed:
use snake_case namespaces for moves inplementation and tests

kilohsakul commented on include/mata/nfa/delta.hh:
```diff
@@ -121,64 +121,105 @@
+     *
+     * It iterates over pairs (symbol, target) for the given @c StatePost.
      */
-    struct moves_const_iterator {
```
Btw, did we want to make state post with a member instead of inheritance from OrdVector?

Adda0 replied:
Yes. We will refactor `StatePost` to have a member `symbol_posts` of type `OrdVector` instead of having it inherit from `Ordvector` for more explicit code which does not utilize more advanced features of C++ (inheritance of classes). It might be easier for students to read and understand the code with members instead of inheritance.

kilohsakul commented on include/mata/nfa/delta.hh:
```diff
@@ -366,68 +409,75 @@
-        bool is_end_;
-        Transition transition_{};
-
+    class Transitions {
```
Hm, I am still wandering whether having all this code just to avoid one for loop when printing out automaton is worth it ... I know that it is stupid to say this after it is done.

Adda0 replied:
In that case, what about the class `Transition` and functions accepting this class? We would have no way to create an instance of `Transition`, so then the whole class would be pointless and could be removed, too, no?

jurajsic replied:
I think having an iterator over all transitions can be useful in some other contexts than printing out automaton. For example if I want to check if automaton accepts only one symbol words, I can iterate all transitions and check if source state is initial and target is final (I feel like we did something like that in Noodler).

kilohsakul replied:
ok, lets keep it

Adda0 committed:
Move Moves and Transitions and its iterators outside their parent classes

Adda0 committed:
Set Delta::operator= to be default implemented again

It could not be default implemented when we had transitions as a
member of Delta. When we no longer have that, it can be default
implemented again using the member-wise copy assignment.

kilohsakul commented on include/mata/nfa/delta.hh:
```diff
@@ -121,64 +121,105 @@
+        class const_iterator {
+        private:
+            const StatePost* state_post_{ nullptr };
+            StatePost::const_iterator state_post_it_{};
```
Names of state_post_it and symbol_post_it, may it be better to name them by to what they are dereferenced to instead of what class owns them? Maybe something as symbol_it and target_it? I am thinking about this since I find the code with the current names a little hard to read.

Adda0 replied:
You either name the iterators after whose iterator it is (as in this case) or over what data the iterator iterates (in this case, it would be `symbol_post_it` iterating over symbol posts). I am fine with either, and I think both approaches are used equally. If you both think it would be better, I am all for changing the name as you suggest.

Adda0 replied:
It was decided that we will name iterators as `<the_name_of_the_object_we_get_by_dereferencing_the_iterator>_it`. We can slowly begin renaming all existing iterator variables to adhere to this convention.

For example, for const iterator over `StatePost`, we get the following variable name:
```cpp
StatePost::const_iterator symbol_post_it = nfa.delta.state_post(4).begin();
```

Adda0 replied:
For further discussion, please refer to https://github.com/VeriFIT/mata/issues/313.

jurajsic commented on include/mata/nfa/delta.hh:
```diff
@@ -444,7 +494,9 @@ public:
      */
     static StatePost::const_iterator epsilon_symbol_posts(const StatePost& state_post, Symbol epsilon = EPSILON);
-}; // struct Delta.
+private:
+    std::vector<StatePost> state_posts_;
+}; // class Delta.
```
Why was this moved to the end of class definition?

Adda0 replied:
I needed more private members earlier (which are not needed now), which had to be defined only after their types have been defined in the Delta class. Therefore, the private members list had to be at the end of the class definition. In the end, only the original member remained, but private members (be it member variables or functions), should not be the first thing you see when you are looking at a class interface, anyway. Henceforth, I decided to keep this change as it is. It helps with readability: the user of Delta does not need to bother themselves with the knowledge that internally, there is some vector of state posts in `Delta`, which they cannot access or do anything with, anyway.

Adda0 committed:
Group function definitions

Adda0 committed:
Refactor moves public interface

Adda0 committed:
Change iterate parameter to seek first symbol direction

Adda0 committed:
Improve comments

Adda0 committed:
Specify iteration over only epsilons with bool flag

Adda0 commented:
>  I would rename const_iterator to ConstIterator to match the convention.

For renaming custom iterators, I have created a separate issue #312. I will keep the current naming convention in this PR. We may rename all custom iterators at once later on when resolving #312.

Adda0 commented:
All the issues and suggestions have been addressed. If you want to add anything more to this PR, let me know. Otherwise, when @kilohsakul reviews the code and his issues are resolved, the PR should be finished and ready to be merged.

Adda0 committed:
Move logic getting begin and end symbol post iterators to moves_...() functions

Adda0 committed:
Simplify last symbol lookup

Adda0 committed:
Add more edge case tests

Adda0 commented:
@kilohsakul What do you think about the latest changes, moving the code handling getting iterators to symbol posts to iterate over to functions `moves()`, `moves_epsilons()` and `moves_symbols()`? Is there anything more you would like to change, or do you approve the PR?

kilohsakul commented on src/nfa/delta.cc:
```diff
@@ -340,82 +343,172 @@ void Delta::defragment(const BoolVector& is_staying, const std::vector<State>& r
+    return *this;
+}
+
+StatePost::Moves StatePost::moves(
```
Good, causes no brain demage now, nice.

Adda0 commented:
PR has been rebased onto the current devel with refactored CMake. Once GAs finish, the PR will be merged.
